### ISSUE TITLE
Gradient not registered in autograph mode with tf.data.Dataset loop

### ISSUE THREAD

**Issue by AdrienCorenflos** [tensorflow/tensorflow#38453]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): True
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or
binary): PIP
- TensorFlow version (use command below): v2.1.0-rc2-17-ge5bf8de410 2.1.0

**Describe the current behavior**
In autograph mode, gradient is not registered properly (propagates None) if function value depends on dataset values

**Describe the expected behavior**
Gradient in autograph should match gradient in eager mode

**Standalone code to reproduce the issue** 
```python
import tensorflow as tf

dataset = tf.data.Dataset.from_tensor_slices([2., 3.])

def fun(x):
    for data in dataset:
        x = x * data
    return x

decorated_fun = tf.function(fun)

a = tf.constant(1.)
with tf.GradientTape(persistent=True) as tape:
    tape.watch(a)
    b = fun(a)
    c = decorated_fun(a)
print(tape.gradient(b, a)) # prints 6. as expected
print(tape.gradient(c, a)) # prints None

```

**Comment by AdrienCorenflos** [comment]
Also note that the dataset example in the [official performance doc](https://www.tensorflow.org/tutorials/customization/performance) wouldn't propagate the gradient either.

```python
@tf.function
def train(a, dataset):
    loss = tf.constant(0.)
    for x, y in dataset:
        loss += a * tf.abs(y - x) # Some dummy computation.
    return loss

data = [(1., 1.)] * 10

dataset = tf.data.Dataset.from_generator(lambda: data, (tf.float32, tf.float32))

a_tensor = tf.constant(1.)
with tf.GradientTape(persistent=True) as tape:
    tape.watch(a_tensor)
    list_val = train(a_tensor, data)
    dataset_val = train(a_tensor, dataset)
print(tape.gradient(list_val, a_tensor)) # prints 0.
print(tape.gradient(dataset_val, a_tensor)) # prints None
```

**Comment by Saduf2019** [comment]
i am able to replicate the code shared, please find the [gist here](https://colab.sandbox.google.com/gist/Saduf2019/3a8f938d5bb54c5e2ce7422ab0d4c88b/38453.ipynb)

**Comment by AdrienCorenflos** [comment]
In the meantime I have found a workaround for the above by going a bit lower level and implementing the while_loop myself. I am posting it here in case it helps tf maintainers to figure out what the bug is related to.

```python

import tensorflow as tf

this_dataset = tf.data.Dataset.from_tensor_slices([2., 3.])

def fun(x, dataset):
    for data in dataset:
        x = x * data
    return x

def low_level_fun(x, dataset, n):
    iterator = iter(dataset)
    i0 = tf.constant(0)
    
    def body(val, i):
        next_data = iterator.get_next()
        return val * next_data, i+1
    
    def cond(_val, i):
        return i < n
    
    res, _ = tf.while_loop(cond, body, [x, i0], back_prop=True)
    
    return res

decorated_fun = tf.function(fun)
decorated_low_level_fun = tf.function(low_level_fun)

a = tf.constant(1.)
with tf.GradientTape(persistent=True) as tape:
    tape.watch(a)
    b = fun(a, this_dataset)
    c = decorated_fun(a, this_dataset)
    d = low_level_fun(a, this_dataset, 2)
    e = decorated_low_level_fun(a, this_dataset, 2)
    
print(tape.gradient(b, a)) # prints 6. as expected
print(tape.gradient(c, a)) # prints None
print(tape.gradient(d, a)) # prints 6. as expected
print(tape.gradient(e, a)) # prints 6. as expected

```

**Comment by jsimsa** [comment]
tf.data does not support gradients.

**Comment by jsimsa** [comment]
The intended use for tf.data is to perform the (non-trainable) [ETL](https://en.wikipedia.org/wiki/Extract,_transform,_load) for data used by your ML workload.

**Comment by AdrienCorenflos** [comment]
OK so what is the recommended interface for building a loss function depending on data (basically something similar to the official doc for performance)?

**Comment by alextp** [comment]
Think of tf.data as a tool for data loading into tensorflow, not for data processing as part of TF.

So we recommend you don't use tf.data to transform a piece of data you already have loaded in memory, as it's not really designed for that. Instead use your own TF loops (which should be relatively straightforward to write with autograph) to do your mapping, filtering, batching (which is just tf.stack), etc, all of which are differentiable operations.

**Comment by alextp** [comment]
Nevermind, I think there's an actual bug here where autograph is generating loops which break autodiff sometimes.

**Comment by mdanatg** [comment]
@AdrienCorenflos `for data in iter(dataset):` will generate gradients as expected. With `iter`, autograph will generate code very similar to what you wrote. Without, it uses reduce which blocks gradients. We'll look into making them behave consistently.

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tf Nightly 2.6.0-dev20210525, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/919a01ca7bab6405ddb543a0fb8ce484/38453.ipynb). Thanks!

**Comment by tilakrayal** [comment]
Was able to reproduce your issue in Tf v2.7 please find the gist [here](https://colab.sandbox.google.com/gist/tilakrayal/92a721fc3289d9b9aa6c704c593cb717/untitled157.ipynb). Thanks!

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
